### PR TITLE
Major diffcalc overhaul and skill rewrite/addition of `Rhythm` and `Reading` within osu!taiko

### PR DESCRIPTION
**Pull request by Lawtrohux**
# Rewrite of Rhythm within osu!taiko

## Ratio Difficulty
In this rework, rhythm difficulty is based on the fact that more complex interval ratios are rhythmically more difficult. In general, we can define more complex ratios as fractions that have larger denominators in their simplified form. For example, 3/5 is more difficult than 1/2. With these assumptions, we construct a sine wave-based function that penalises the specified denominator.

---

## Rhythm Changes
- Introduction of three new rhythm data classes:
  - `EvenHitObjects`
  - `EvenPatterns`
  - `EvenRhythm`
- Intervals are accessed via the `IHasInterval` interface.

### Move Rhythm into an Evaluator for per-object difficulty.
- Penalises consistent intervals within rhythm.
- Gives bonuses to near-1 ratio's, and penalises very-near-1 ratio's.
- Penalises exceedingly long gaps between awkward rhythm (primarily found in low SR difficulties).

---

## Colour Changes
- Same interval colour changes are now penalised based on their `deltatime` variance. Same `deltatime`, and the more it accumlates, the more of a penalty is applied.

---

# Introduction of the Reading Skill within osu!taiko

## Reading Evaluator
Calculates the influence of higher slider velocities on hitobject difficulty. The bonus is determined based on the EffectiveBPM, shifting within a defined range of a sigmoid function between the upper and lower boundaries to reflect how increased slider velocity impacts difficulty.

Low SV does contain considerations within this build. These are pretty minimal, but at most the more dense a map is, the harder its difficulty will be.

- Drumrolls and Swells are exempt from difficulty calculation, as they are not required to be hit.
- Rhythm actually decay's based on time, and not notes.

## EffectiveBPM Preprocessor
Calculates and sets the effective BPM and slider velocity for each note object, considering clock rate and scroll speed. This is map based, as both its base slider velocity and its scroll speed are defined differently, is determined by accessing ControlPoints for timing.

# Difficulty Calculation Changes
Introduces `PatternPenalties`, which calculates the combined penalty based on the relationship between skills. Lower skill values are penalised more heavily relative to predefined thresholds and their interaction with the opposing skill rating.

- Looks at the relationship between colour and rhythm, rhythm below a value of 2.5 with high colour difficulty gets penalised. This is done in a way that doesn't affect most normal maps, only one's that utilised abuse of the colour system previously. Similarly, maps with low colour and high rhythm are penalised too, this is shown mostly to affect lower star rating's as rhythm has been underweight for quite some time now.

#### Introduces `countdifficultstrains` to taiko calculation  
  These are used in two instances:
  - One as a length bonus to stamina, which kicks in at a count of 1250, with a slow increase to 1350 and significantly past then. It is capped past 2000 difficult strains.
  - Rhythm strains below 50 are slowly scaled down to ensure that awkward intervals with a low length are not overbuffed.

### New Attributes
- Skill-based difficult strain
- `SimplePatternFactor` (PatternPenalties)
- `ObjectDensity`
- `Reading`

---

## Performance Points Changes
Performance Points changes were kept to a minimal in order to fix how star rating functions currently and be able to test within an environment without too many parameters.

- Removed fixed HR Difficulty Performance Point multipliers , this is now handled only within Star Rating Calculation and Accuracy side of Performance Points.

- Adjusted statistical accuracy weight from `300` to `400`

- Improved the accuracy "floor" for a greater range of scaling between lower accuracies

---

## Where can I view changes? 
For a dev-based writeup on the Rhythm and Reading Skills, these are viewable here:  
[Dev Writeup](https://docs.google.com/document/d/1-k96U-3-77CSyWNbXxa3cf2AXeOIxVh8yoXowlIYc5g/edit?usp=sharing)

For a very general writeup meant for end users, view these here:  
[General Writeup](https://www.ltca.pro/osu!taiko/ppv3)

For testing of values, this can be viewed on huismetbenen:  
[huismetbenen](https://pp.huismetbenen.nl/rankings/players/taiko-ppv3)

## Some already present issues
Due to strain chunking, rate changing some maps can cause values to go up or down depending on very small rate changes, this is also present within the `MarginOfError` with rhythm, as slight interval changes (outside of 5ms) move pattern grouping. This doesn't contribute to any openings for abuse. This is something outside of the constraint of what this pr aims to achieve, and as rate changes stay unranked, depends on the strain system.

### CODE REVIEW

**Comment by Lawtrohux** [comment]
@smoogipoo could i please get a smoogisheet for this branch? converts off/ranked-only if possible.

**Comment by Lawtrohux** [comment]
@ppy/taiko-pp-commitee after a couple weeks with the community and balancing, ready for final review.

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/IHasInterval.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Rhythm
+{
+    /// <summary>
+    /// The interface for hitobjects that provide an interval value,
```
Comment (YaniFR) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/IHasInterval.cs:7]: Replace the comma with a point ?

File: osu.Game.Rulesets.Taiko/Difficulty/Evaluators/RhythmEvaluator.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            // If there are fewer than 4 valid intervals, skip the consistency check.
+            if (intervals.Count < interval_count)
+                return 1.0; // No penalty applied if there isn't enough data.
+
+            for (int i = 0; i < intervals.Count; i++)
+            {
+                for (int j = i + 1; j < intervals.Count; j++)
+                {
+                    double ratio = intervals[i]!.Value / intervals[j]!.Value;
+                    if (Math.Abs(1 - ratio) <= threshold) // If any two intervals are similar, apply penalty.
```
Comment (YaniFR) [inline osu.Game.Rulesets.Taiko/Difficulty/Evaluators/RhythmEvaluator.cs:75]: "apply a penalty" instead of "apply penalty"

**Review by YaniFR — COMMENTED** [review]
Sounds good for me except for the 2 little thing i mentionned,
some comments are missing point too but i don't think its too relevant to be considered

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyHitObject.cs [hunk 3]
```diff
@@ -71,7 +87,7 @@ public TaikoDifficultyHitObject(HitObject hitObject, HitObject lastObject, HitOb
 
             // Create the Colour object, its properties should be filled in by TaikoDifficultyPreprocessor
```
Comment (kadenbruh) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyHitObject.cs:88]: Make sure this comment also includes Rhythm

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs [hunk 4]
```diff
@@ -29,15 +29,13 @@ public Reading(Mod[] mods)
 
         protected override double StrainValueOf(DifficultyHitObject current)
         {
-            if (current is not TaikoDifficultyHitObject taikoObject)
-                return 0.0;
-
-            // Drum Rolls and Swells are exempt.
             if (current.BaseObject is not Hit)
```
Comment (kadenbruh) [inline osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs:30]: Add the comment from before back as this still applies. 

// Drum Rolls and Swells are exempt.

Comment (kadenbruh) [inline osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs:30]: Add the comment from before back as this still applies.

// Drum Rolls and Swells are exempt.

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyHitObject.cs (hunk 3 above)
Comment (kadenbruh) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyHitObject.cs:88]: This comment needs to include Rhythm object.

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        private double currentStrain;
+
+        public Reading(Mod[] mods)
+            : base(mods)
+        {
+        }
+
+        protected override double StrainValueOf(DifficultyHitObject current)
+        {
+            if (current.BaseObject is not Hit)
```
Comment (kadenbruh) [inline osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs:30]: Add the comment from before back as this still applies.

// Drum Rolls and Swells are exempt.

**Review by kadencebruh — CHANGES_REQUESTED** [review]
Small comment additions that can be made for better clarity. Feel free to add them if you wish.

**Comment by Lawtrohux** [comment]
Closing this on the basis of splitting out pr's, as we now have the seperate branch.

**Pull request closed without merging**